spec/requests/omniauth_callbacks_spec.rb

```diff
@@ -0,0 +1,382 @@
+# frozen_string_literal: true
+
+describe 'OmniauthCallbacks', type: :request do
+  let(:oauth_email) { 'example@xyze.it' }
+  let(:oauth_token) { 'abcdefg12345' }
+
+  def set_omniauth_auth_env(provider:)
+    auth_hash = OmniAuth::AuthHash.new(
+      provider: provider,
+      uid: '1234',
+      info: {
+        email: oauth_email,
+        name: 'Test User',
+        image: 'http://example.com/image.jpg'
+      },
+      credentials: {
+        token: oauth_token,
+        refresh_token: '12345abcdefg',
+        expires_at: DateTime.now.in_time_zone
+      }
+    )
+
+    Rails.application.env_config['omniauth.auth'] = auth_hash
+  end
+
+  def set_devise_mapping_env
+    Rails.application.env_config['devise.mapping'] = Devise.mappings[:user]
+  end
+
+  def set_invitation_token_env(invitee)
+    Rails.application.env_config['omniauth.params'] = { 'invitation_token' => invitee.invitation_token }
+  end
+
+  def set_avatar_env(avatar_image_url)
+    Rails.application.env_config['omniauth.auth']['info']['image'] = avatar_image_url
+  end
+
+  def invite_user(email)
+    inviter = create(:user)
+    invitee = User.invite!({ email: email }, inviter)
+    set_invitation_token_env(invitee)
+  end
+
+  describe 'GET #google_oauth2' do
+    before do
+      set_devise_mapping_env
+      set_omniauth_auth_env(provider: 'google_oauth2')
+    end
+
+    context 'when google_oauth2 email doesnt exist in the system' do
+      it 'creates user with info in google_oauth2' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+        user = User.find_by(email: oauth_email)
+        expect(user).to be_present
+      end
+
+      it 'signs in and redirect to root' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+        expect(response).to redirect_to root_path
+      end
+
+      it 'enables oauth on user' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+        user = User.find_by(email: oauth_email)
+        expect(user.oauth_enabled?).to eq true
+      end
+    end
+
+    context 'when google_oauth2 email already exist in the system' do
+      let!(:user) { create(:user, email: oauth_email, token: nil) }
+
+      it 'updates the user with google_oauth2 credentials' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+        user = User.find_by(email: oauth_email)
+        expect(user.token).to eq oauth_token
```

Comment:
I'm curious, but are we expecting the `oauth_token` to change here? If so, I think we can use an **assume** step to make sure it was set to something else. Just wondering what it's trying to update here.

```suggestion
        user = User.find_by(email: oauth_email)
        expect(user.token).not_to eq oauth_token
        get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
        expect(user.reload.token).to eq oauth_token
```

Reply:
I made the test a bit cleared and moved user creation from the `let!` to the `it` scope. What do you think?

Reply:
That's fine with me to be more explicit going by DAMP vs DRY principles. `let!` would be more DRY while putting `create` in `it` would be more DAMP. 

There's a good summary of the topic on: https://stackoverflow.com/a/11837973/9463774

Pretty much when to favor DRY or DAMP pretty much kind of depends how you want to balance out readability vs duplication.